### PR TITLE
feat: support general capacity restrict for cloud-side resoure contro…

### PR DESCRIPTION
**Pull request by MrPresent-Han**
related: #29844

### CODE REVIEW

**Comment by MrPresent-Han** [comment]
rerun ut

**Comment by MrPresent-Han** [comment]
rerun ut

File: internal/rootcoord/constrant.go [hunk 1]
```diff
@@ -16,10 +16,57 @@
...
+		addedNum += newParNum
+	}
+
+	databases, err := core.meta.ListDatabases(ctx, ts)
+	if err != nil {
+		return fmt.Errorf("failed to list databases when checking general capacity")
+	}
+	var generalNum int64 = 0
+	for _, database := range databases {
+		collections, err := core.meta.ListCollections(ctx, database.Name, ts, false)
```
Comment (czs007) [inline internal/rootcoord/constrant.go:56]: Why not simply enumerate the collections that are available?

File: pkg/util/paramtable/component_param.go [hunk 2]
```diff
@@ -941,6 +942,21 @@ func (p *rootCoordConfig) init(base *BaseTable) {
 		Export:       true,
 	}
 	p.MaxDatabaseNum.Init(base.mgr)
+
+	p.MaxGeneralCapacity = ParamItem{
+		Key:          "rootCoord.maxGeneralCapacity",
+		Version:      "2.3.0",
```
Comment (czs007) [inline pkg/util/paramtable/component_param.go:948]: 2.3.5

File: internal/rootcoord/create_collection_task.go [hunk 3]
```diff
@@ -66,18 +66,15 @@ func (t *createCollectionTask) validate() error {
...
 	shardsNum := t.Req.GetShardsNum()
 
 	cfgMaxShardNum := Params.RootCoordCfg.DmlChannelNum.GetAsInt32()
 	if shardsNum > cfgMaxShardNum {
 		return fmt.Errorf("shard num (%d) exceeds max configuration (%d)", shardsNum, cfgMaxShardNum)
 	}
 
-	cfgShardLimit := Params.ProxyCfg.MaxShardNum.GetAsInt32()
-	if shardsNum > cfgShardLimit {
-		return fmt.Errorf("shard num (%d) exceeds system limit (%d)", shardsNum, cfgShardLimit)
```
Comment (czs007) [inline internal/rootcoord/create_collection_task.go:78]: I believe it is still necessary to maintain this logic. It restricts the number of shards for a single table from exceeding the maximum configured shard count.

**Comment by MrPresent-Han** [comment]
rerun ut

**Comment by MrPresent-Han** [comment]
rerun ut

File: pkg/util/paramtable/component_param.go [hunk 4]
```diff
@@ -862,6 +862,7 @@ type rootCoordConfig struct {
 	ImportTaskSubPath           ParamItem `refreshable:"true"`
 	EnableActiveStandby         ParamItem `refreshable:"false"`
 	MaxDatabaseNum              ParamItem `refreshable:"false"`
+	MaxGeneralCapacity          ParamItem `refreshable:"false"`
```
Comment (longjiquan) [inline pkg/util/paramtable/component_param.go:865]: Maybe we should make this parameter refreshable.

**Review by longjiquan — CHANGES_REQUESTED** [review]

File: pkg/util/paramtable/component_param.go (hunk 4 above)
Comment (MrPresent-Han) [reply pkg/util/paramtable/component_param.go]: changed

**Comment by MrPresent-Han** [comment]
rerun ut

**Comment by MrPresent-Han** [comment]
rerun ut

**Comment by MrPresent-Han** [comment]
rerun ut

**Pull request merged**